### ISSUE TITLE
BUG fetch_diabetes_hospital does not have as_frame argument

### ISSUE THREAD

**Issue by MiroDudik** [fairlearn/fairlearn#1185]
#### Describe the bug
The function [`fairlearn.datasets.fetch_diabetes_hospital`](https://fairlearn.org/main/api_reference/fairlearn.datasets.html#fairlearn.datasets.fetch_diabetes_hospital) does not have `as_frame` argument, unlike all other fetch functions in the datasets module.

#### Suggested fix
Add `as_frame` argument to `fetch_diabetes_hospital`. We could possibly also rethink the default behavior of the fetch functions. My sense is that defaulting to `as_frame=False` is actually the opposite of what I'd like whenever I use the fetch functions (which is basically just in educational resources).

**Comment by shimst3r** [comment]
I'd like to tackle this so I can continue working on #1182 🤓

**Comment by shimst3r** [comment]
@MiroDudik given that `as_frame=False` is the default for the other fetch functions, but `fetch_diabetes_hospital` has an implicit `as_frame=True` (because the current logic assumes the return value to be a data frame), I would propose to implement the solution for this issue using `as_frame=True` (to avoid breaking changes for people who haven't set the function argument explicitly).

Regarding the other fetch functions:
- In my opinion, **not** returning a data frame as the default is sensible. It's a more complex data structure than a `numpy.ndarray` and hence *feels* like setting an additional flag is okay.
- Would you say that change needs a dedicated issue with proper change management (potentially even a logger info with a grace period)?

**Comment by MiroDudik** [comment]
I think that we should have a consistent default across all of `fetch_*` functions.

But you are definitely right that we shouldn't change the default behavior on `fetch_diabetes_hospital`, where it's effectively `True`, or on all other functions, where it is `False`, without some transition period. Once we decide what the new default should be, we should make sure that those functions where the default will change issue a warning when called without specifying `as_frame`. For example, the warning would be something like:

>You have not specified the value for `as_frame`, which currently defaults to `True`. Starting with version 0.10.0, the default is changing to `False`. Specify the value of `as_frame` to suppress the warning.

I actually think that the more sensible default value is `True` for all `fetch_*` functions. They are really just meant for educational purposes, so returning a DataFrame seems more appropriate. I reviewed our documentation and examples. We are currently using three datasets: `fetch_adult(as_frame=True)`, `fetch_boston(as_frame=True)`, and `fetch_diabetes_hospital()`. So we never use `as_frame=False` (either explicitly or via default). The default `True` would be also closer to [sklearn default for fetch_openml](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_openml.html).

@shimst3r , what do you think?

@fairlearn/fairlearn-maintainers : do you have thoughts on the default value for `as_frame` in all `fetch_*` functions?

**Comment by shimst3r** [comment]
> I think that we should have a consistent default across all of fetch_* functions.
I totally agree. 👍 

> we should make sure that those functions where the default will change issue a warning when called without specifying `as_frame`
@MiroDudik I like that idea. What do you think of using a decorator similar to this one:
```python
import logging
from functools import wraps

logger = logging.getLogger(__name__)


def use_as_frame_explicitly(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        if "as_frame" not in kwargs:
            msg = (
                "You have not specified the value for as_frame, which currently "
                "defaults to True. Starting with version 0.10.0, the default is "
                "changing to False. Specify the value of as_frame to suppress "
                "the warning."
            )
            logger.warning(msg)
        return f(*args, **kwargs)

    return wrapper
```

**Comment by MiroDudik** [comment]
Was the idea to define the decorator once and then use it in multiple fetch functions? I might be okay with that, but I find decorators a bit hard to follow, so I would prefer something like this:
https://github.com/fairlearn/fairlearn/blob/715eb13d45593decf37fdb0058182e2da93ae94e/fairlearn/postprocessing/_threshold_optimizer.py#L291-L301
There, we change the default value to 'deprecated' and then have an if statement to produce the warning (and also appropriately set the value of the argument to get the old behavior).

We should probably modularize the above into a function ``check_as_frame(as_frame, old_default_value)`` that is reused across all fetch_functions and called like this:
```pyhon
as_frame = check_as_frame(as_frame, True)
```

@fairlearn/fairlearn-maintainers -- thoughts on the default value for `as_frame` going forward?

**Comment by adrinjalali** [comment]
I would like us to be consistent, even if it breaks backward compatibility here (we can do it through a deprecation cycle, but I don't mind not doing that here since these are used only for education purposes)

sklearn has `as_frame=False` everywhere, we could follow that, or have it everywhere as `True` for the default. I don't mind either way.

I don't think we need a decorator here, but I also don't mind that. Seems a bit overkill to me.

**Comment by MiroDudik** [comment]
@adrinjalali -- actually sklearn now defaults to `as_frame='auto'`, which becomes `True` (except in cases when the data set is stored as a sparse matrix, which I don't think is the case for any of our datasets anyway). So also from that side, I think that `True` might be closer to the expected behavior.

I would be fine breaking compatibility in this specific instance without any deprecation cycle... precisely for the reason you mention.

**Comment by shimst3r** [comment]
> I might be okay with that, but I find decorators a bit hard to follow, [...]

I haven't thought of that aspect to be honest, guess it was a matter of premature optimisation. 😄 If that's the prevailing opinion, I commit to implementing it as a regular function within the fetch functions' bodies using a new default argument. 👍

**Comment by romanlutz** [comment]
@MiroDudik shall we leave this issue open for the actual dataset update in OpenML or shall we create a new issue for that?
I'm somewhat in favor of the latter, but I'm fine with it either way.

Huge thanks to @shimst3r for taking care of #1194 including all our historical baggage around datasets 🤣

**Comment by MiroDudik** [comment]
@romanlutz -- yes. Let's create a new issue to change the default to `as_frame=True` across all `fetch_*`? It seems there's a consensus that we don't need a deprecation cycle (https://github.com/fairlearn/fairlearn/issues/1185#issuecomment-1403743188)

**Comment by romanlutz** [comment]
Perfect. I opened #1206 for this and will closed this issue since @shimst3r 's PR #1194 is merged.
